test/blackbox/common/DDSBlackboxTestsReadConditions.cpp

```diff
@@ -0,0 +1,109 @@
+using eprosima::fastrtps::types::ReturnCode_t;
+
+TEST(ReadConditions, read_conditions_management)
+{
+    PubSubReader<HelloWorldPubSubType> manager(TEST_TOPIC_NAME);
+    manager.init();
+    ASSERT_TRUE(manager.isInitialized());
+    DataReader& reader = manager.get_native_reader();
+
+    // Condition masks
+    SampleStateMask sample_states = 0;
+    ViewStateMask view_states = 0;
+    InstanceStateMask instance_states = 0;
+
+    // Create and destroy testing
+
+    // 1- cannot create a ReadConditon that cannot be triggered
+    ReadCondition* cond = reader.create_readcondition( sample_states, view_states, instance_states);
+    EXPECT_EQ(cond, nullptr);
+
+    // 2- create a ReadCondition and destroy it
+    sample_states = ANY_SAMPLE_STATE;
+    cond = reader.create_readcondition( sample_states, view_states, instance_states);
+    EXPECT_NE(cond, nullptr);
+    ReturnCode_t res = reader.delete_readcondition(cond);
+    EXPECT_EQ(res, ReturnCode_t::RETCODE_OK);
+
+    // 3- Create several ReadConditions associated to the same masks (share implementation)
+    std::forward_list<ReadCondition*> conds;
+
+    for (int i = 0; i < 10 ; ++i )
+    {
+        conds.push_front(reader.create_readcondition( sample_states, view_states, instance_states));
+    }
+
+    for (ReadCondition* c : conds)
+    {
+        EXPECT_EQ(reader.delete_readcondition(c), ReturnCode_t::RETCODE_OK);
+    }
+    conds.clear();
+
+    // 4- Create several ReadConditions associated to different same masks
+    sample_states = 0;
+    view_states = 0;
+    instance_states = 0;
+
+    for (int i = 0; i < 10 ; ++i )
+    {
+        conds.push_front(reader.create_readcondition( ++sample_states, ++view_states, ++instance_states));
+    }
+
+    for (ReadCondition* c : conds)
+    {
+        EXPECT_EQ(reader.delete_readcondition(c), ReturnCode_t::RETCODE_OK);
+    }
+    conds.clear();
+
+    // 5- Create several ReadConditions and destroy them using delete_contained_entities
+    sample_states = 0;
+    view_states = 0;
+    instance_states = 0;
+
+    for (int i = 0; i < 10 ; ++i )
+    {
+        conds.push_front(reader.create_readcondition( ++sample_states, ++view_states, ++instance_states));
+    }
+
+    EXPECT_EQ(reader.delete_contained_entities(), ReturnCode_t::RETCODE_OK);
+    conds.clear();
+
+    // 6- Check a DataReader only handles its own ReadConditions
+    PubSubReader<HelloWorldPubSubType> another_manager(TEST_TOPIC_NAME);
+    another_manager.init();
+    ASSERT_TRUE(another_manager.isInitialized());
+    DataReader& another_reader = another_manager.get_native_reader();
+    cond = another_reader.create_readcondition(sample_states, view_states, instance_states);
+    EXPECT_NE(cond, nullptr);
+    EXPECT_EQ(reader.delete_readcondition(cond), ReturnCode_t::RETCODE_PRECONDITION_NOT_MET);
+
+    // 7- Check the DataReader cannot be deleted with outstanding conditions
+    EXPECT_EQ(another_manager.get_native_subscriber().delete_contained_entities(),
```

Comment:
I don't think the existence of ReadConditions should cause the delete_contained_entities to fail.

Reply:
a) `ReturnCode_t DataReaderImpl::delete_contained_entities()`

My implementation clears all ReadConditions (we can do that because Company implementation unlinks the WaitSets on Condition destruction.

According to the DDS standard 1.4 in 2.2.2.2.1.18 & 2.2.2.5.3.31 delete_contained_entities() will be used recursively and remove all API created objects (including Read & QueryConditions object created by the DataReader).

The operation will return PRECONDITION_NOT_MET if the any of the contained entities is in a state where it cannot be deleted. This is quite vague but seems our DataReaderImpl fulfills the spirit.

b) `ReturnCode_t SubscriberImpl::delete_contained_entities()`

The implementation tries to destroy all DataReaders checking DataReaderImpl::can_be_deleted().
My can_be_deleted() implementation takes into account outstanding ReadConditions (formerly only loans).
This doesn't match the DDS standard 2.2.2.5.2.14 behaviour.

c) `ReturnCode_t SubscriberImpl::delete_datareader(const DataReader* reader)`
    The implementation tries to destroy all the readers querying can_be_deleted() too.
    This matches standard 2.2.2.5.2.6 description (any outstanding ReadConditions will lead to failure).

It's clear the issue is b). b) and c) are supposed to behave differently but are calling the same DataReaderImpl::can_be_deleted()
method. The solution is make clear to the method which function is calling it by using a parameter as:
```c++
bool DataReaderImpl::can_be_deleted(bool recursive = true) const
```